File: nivel3/generala_nivel3/generala.py
import random
from copy import copy
def generala(a,b,c,d,e):
    valores = []
    numeros = [a,b,c,d,e]
    #? Suma de los dados
    puntos = [0,0,0,0,0,0]
    for numero in numeros:
        puntos[numero -1] += numero
        
    for punto in puntos:
        valores.append(punto)
    
    #? Escalera
    
    cantidades = []
    for i in range(1,7):
        cantidad = numeros.count(i)
        cantidades.append(cantidad)

    cantidades1 = copy(cantidades)
    cantidades1.pop(5)
    cantidades2 = copy(cantidades)
    cantidades2.pop(0)
    cantidades3 = copy(cantidades)
    cantidades3.pop(1)
    casos = (cantidades1,cantidades2,cantidades3)
    escalera = True
     
    for caso in casos:
        if not caso.count(0):
            escalera = True
        else:
            escalera = False
            
    if escalera: valores.append(20)
    else: valores.append(0)
            
        
       
    #? Full y Poker y Generala
    
    dos_iguales = False
    tres_iguales = False
    cuatro_iguales = False
    cinco_iguales = False
    for i,punto in enumerate(puntos):
        i += 1
        if punto == 0:
            continue
        elif punto / 2 == i:
            dos_iguales = True
        elif punto / 3 == i:
            tres_iguales = True
        elif punto / 4 == i:
            cuatro_iguales = True
        elif punto / 5 == i:
            cinco_iguales = True
            
    
    if dos_iguales and tres_iguales: valores.append(30)
    else: valores.append(0)
    
    if cuatro_iguales or cinco_iguales: valores.append(40)
    else: valores.append(0)
    
    if cinco_iguales: valores.append(50)
    else: valores.append(0)

    return valores

print(generala(3,3,3,3,3))
print(generala(1,5,6,3,4))
print(generala(3,2,2,3,2))

File: jurisdiccional/terminados/generala/generala.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int contar(vector<int> numeros, int x)
{
    int contador = 0;
    for (int i = 0; i < numeros.size(); i++)
    {
        if (numeros[i] == x){
            contador++;
        }
    }
    return contador;
}



void generala(int a,int b,int c,int d,int e)
{   
    vector<int> numeros = {a,b,c,d,e};
    vector<int> puntos = {0,0,0,0,0,0};
    vector<int> puntosDePartida(6);

    //Puntos individuales
    for (int i = 0; i < numeros.size(); i++)
    {
        puntos[numeros[i] - 1] += numeros[i];
        puntosDePartida[numeros[i] - 1] += numeros[i];
    }

    //Escalera
    vector<int> numerosCopia = numeros;
    sort(numerosCopia.begin(),numerosCopia.end());
    vector<int> cantidades(6);
    for (int i = 1; i <= 6;i++)
    {
        int cantidad = contar(numerosCopia,i);
        cantidades[i-1] = cantidad;
    }



    vector<vector<int>> casos = {{1,1,1,1,1,0},{1,0,1,1,1,1},{0,1,1,1,1,1}};
    int escalera = 0;

    for (int i = 0; i < 3; i++)
    {
        if (casos[i] == cantidades)
        {
            escalera = 20;
            break;
        }
    }
    puntosDePartida.push_back(escalera);

    //Full,Poker,Generala
    bool dosVeces = false;
    bool tresVeces = false;
    bool cuatroVeces = false;
    bool cincoVeces = false;
    for (int i = 1; i <= 6; i++)
    {   
        int cantidad = puntos[i - 1] / i;
        if (cantidad == 2)
        {
            dosVeces = true;
        }
        if (cantidad == 3)
        {
            tresVeces = true;
        }
        if (cantidad >= 4)
        {
            cuatroVeces = true;
        }
        if (cantidad == 5)
        {
            cincoVeces = true;
        }
        
    }

    if (dosVeces && tresVeces) puntosDePartida.push_back(30); else puntosDePartida.push_back(0);

    if (cuatroVeces || cincoVeces) puntosDePartida.push_back(40); else puntosDePartida.push_back(0);

    if (cincoVeces) puntosDePartida.push_back(50); else puntosDePartida.push_back(0);

    for (int i = 0; i < puntosDePartida.size();i++)
    {
        if (i < puntosDePartida.size() - 1)
        {
        cout << puntosDePartida[i] << ", ";
        }
        else cout << puntosDePartida[i] << ".";
    }
    
}

int main()
{
    generala(3,3,3,3,3);
    cout << endl;
    generala(1,3,4,6,5);
    cout << endl;
    generala(3,2,2,3,2);
}
